File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/playlist/PlaylistBulkRemovePatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to Morphe contributions.
 */

package app.morphe.extension.youtube.patches.playlist;

import android.support.v7.widget.RecyclerView;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.youtube.settings.Settings;

/**
 * Lets several videos be removed from a playlist at once, instead of swiping and confirming one
 * at a time.
 * <p>
 * A checkbox is drawn over each playlist row and a bar along the bottom removes everything
 * picked in a single request. The checkboxes are drawn rather than added as views because the
 * rows are LithoViews, which refuse added children outright.
 */
public final class PlaylistBulkRemovePatch {

    private PlaylistBulkRemovePatch() {
    }

    /**
     * Injection point. Called for every Litho-backed RecyclerView the app creates, most of which
     * are not playlists. The controller stays dormant until the Litho filter says otherwise.
     */
    public static void onFlyoutMenuCreate(RecyclerView recyclerView) {
        try {
            if (!Settings.PLAYLIST_BULK_REMOVE.get()) {
                return;
            }
            new PlaylistSelectionController(recyclerView).attach();
        } catch (Exception ex) {
            Logger.printException(() -> "onFlyoutMenuCreate failure", ex);
        }
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/layout/playlistbulkremove/PlaylistBulkRemovePatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.patches.youtube.layout.playlistbulkremove

import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.shared.misc.litho.filter.addLithoFilter
import app.morphe.patches.shared.misc.settings.preference.SwitchPreference
import app.morphe.patches.youtube.misc.extension.sharedExtensionPatch
import app.morphe.patches.youtube.misc.litho.filter.lithoFilterPatch
import app.morphe.patches.youtube.misc.recyclerviewtree.addRecyclerViewTreeHook
import app.morphe.patches.youtube.misc.recyclerviewtree.recyclerViewTreeHookPatch
import app.morphe.patches.youtube.misc.settings.PreferenceScreen
import app.morphe.patches.youtube.misc.settings.settingsPatch
import app.morphe.patches.youtube.shared.Constants.COMPATIBILITY_YOUTUBE

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/youtube/patches/playlist/PlaylistBulkRemovePatch;"

private const val EXTENSION_FILTER =
    "Lapp/morphe/extension/youtube/patches/components/PlaylistRowFilter;"

/**
 * Removing several videos from a playlist normally means swiping each row, choosing remove and
 * confirming, one video at a time. This draws a checkbox on each row and removes everything
 * selected in a single request.
 *
 * No new bytecode hook is needed. Two existing ones carry everything:
 *
 * - The RecyclerView tree hook hands over every Litho backed list, playlists included. Its
 *   extension method is named `onFlyoutMenuCreate` only because flyout menus were its first
 *   user; the fingerprint matches the generic `LithoRVSLCBinder` constructor.
 * - A Litho filter identifies which of those lists is a playlist, and reads the playlist id out
 *   of the page header component's buffer. YouTube's own browse id hook was tried first and only
 *   ever reports `FEwhat_to_watch`.
 */
val playlistBulkRemovePatch = bytecodePatch(
    name = "Playlist bulk remove",
    description = "Adds an option to show a checkbox on each video in a playlist, so several can be removed at once.",
) {
    dependsOn(
        sharedExtensionPatch,
        settingsPatch,
        lithoFilterPatch,
        recyclerViewTreeHookPatch,
    )

    compatibleWith(COMPATIBILITY_YOUTUBE)

    execute {
        PreferenceScreen.FEED.addPreferences(
            SwitchPreference("morphe_playlist_bulk_remove"),
        )

        addRecyclerViewTreeHook(EXTENSION_CLASS)
        addLithoFilter(EXTENSION_FILTER)
    }
}
